### ISSUE TITLE
Proposal: Remove value from put on Interactors, Repositories & DataSources

### ISSUE THREAD

**Issue by franmontiel** [mobilejazz/harmony-reference#74]
Replace put(value: T, query: Query) by put (query: Query)

When we want to put an object we can make use of an ObjectQuery.

**Comment by RubenVot** [comment]
I vote YES but I think that @vilanovi had some comments against it. I reserve the right to change my mind.

**Comment by jfilowk** [comment]
https://github.com/mobilejazz/harmony-kotlin/blob/feature/kotlin-multiplatform/common/src/commonMain/kotlin/com.harmony.kotlin/data/datasource/network/GenericNetworkDataSource.kt#L94

Just want to mention that for the new GenericNetworkDataSource we use this value for HTTP PUT operations

**Comment by vilanovi** [comment]
Honestly, I don't think this is a good idea for many reasons.

Having an optional "payload" on put operations is very handy in multiple scenarios:

**Key-Value based Data Sources**

When using KeyQuery to store an object for a "value" (key-value storage based data-sources), the fact of being able to re-use the same Query on both Get & Put methods is just great.

```
// Obtaining the object from a query that implements KeyQuery
let object = myGetInteractor.get(MyQueryWithParams(params))

// Storing an object from a query that implements KeyQuery
myPutInteractor.put(object, MyQueryWithParams(params))
``` 

Without a "value" in the put method, the only way to do this would be by having a custom query that contains the "object" plus returns in the KeyQuery interface the same key as the original one. It is doable? yes, but it requires to duplicate all queries to include a payload.

**CacheRepository**

A similar thing happens with the CacheRepository. When you request an object (or an array of objects) with a `CacheSyncOperation` the same query is going to be used in both Get and Put methods (fetching object from storage, if not available, fetching from network and then PUTting it in storage). If we don't have a "value" in the PUT method, we will need to create a new query.. but... which query is this to be created? The CacheRepsitory wouldn't know.

Probably, we could just have an interface called "PayloadQuery" which adds a generic type T attribute "value" to an existing query, and we would need to require only "PayloadQuery" types in CacheRepository get methods. But then we enter the difficult terrain of handling Generic types in our queries, which doesn't work good in all languages (in Typescript for instance, this wouldn't be possible).

**My opinion**

The "value" in put methods is just a convenience thing: you want to use it? you can. You don't want to use it? just don't. Nobody is forcing you to always pass objects in the put methods, as you can always bundle your changes within a query and use it instead of passing a value object.

It is actually a very good practice to have specific queries for most of your put interactions, as with a query you have to provide an explicit definition of what is happening and which attributes are you changing. So, feel free to use it as much as you want.

However, I'd vote against removing the "value" option, as in some scenarios is very handy and useful and solves the problem of adding a "payload" in a very easy way.

And we haven't even jumped into retro-compatibility terrain, which is not and this would apply a major checkpoint in the usage of harmony in old vs new projects (so you even have to ask yourself if this change is really needed and which are the benefits of it).

If anyone has a strong statement on why we should apply this change, please it's the time to rise the argument.

**Comment by joselufo** [comment]
100% agree with @vilanovi. Besides the problems already exposed by Joan, there is another problem using only `Query` and not using the `value: T?` in the datasources. 

The generics in the `Query`, for example, `ObjectQuery<T>`, are _erased_ at runtime because the generic type information is lost in the runtime. So we can't check the `Query` generic type like `foo as ObjectQuery<User>` because in the runtime is `ObjectQuery<*>`. This is no a good practice and also an inconvenience because we should need to do two different checks (maybe, two `when` operations) just to check the `Query` type and then the value type.

**Comment by franmontiel** [comment]
Agree with @vilanovi & @joselufo. Taking all of that into account I see no clear advantage in changing it.

**Comment by doup** [comment]
**TLDR;**

After writing this, two things:
1. I think what we're lacking standardised ways of doing common things like create & update.
2. I think a reactive state management datasource is feasible without changing all the interfaces (lol, now I want to explore more this)

---

Just for the record, my initial question was if removing the `value` is something we already discussed. I'm not sure either if it's a good idea, I just wanted to know about prior discussions (if they existed), or good reason against it.

I'm not sure about the `CacheRepository`/`KeyValue` argument, wouldn't it be a matter of adding a new Query that extends `KeyValue` that would be used for `put`? You try to `get`, otherwise you write, but using the payload on the Query. But since the `put` query extends the one `get` needs, everything works.

@joselufo I don't understand the generics issue.

What bothers me is that it's confusing when to use just a Query, when a value+Query and when just a value. So you want to create a `Foo` object? These can be some ways I could think of doing this (I'm quite sure I'm not the only one confused by the options):

- `put(null, new CreateFooQuery(...))`: Everything on the Query
- `put(new Foo())`: No ID, so it's an implicit create
- `put(new Foo(id: 10), new CreateFooQuery())`: I'll ignore ID, as the query has an explicit command

Removing the `value`, removes ambiguity. Although, this can be also solved by setting a standard for both create & update (and maybe other common operations).

In the other hand, maybe an argument against removing `value` is that then we would need to map the queries (although, maybe it's just a consequence; and doesn't change what we currently have); or more precisely the payloads. Which, on the other hand maybe is interesting because sometimes is weird to define a Query with some properties that don't match those used in the DataSource[1] (you know, because mappers, but mappers don't apply to queries). Probably it's me that I'm doing it wrong, but sometimes I had the impression that queries could use mappers. Maybe, there isn't any issue there. The data layer knows everything about the domain layer? But the domain layer just knows about the data layer interfaces (repos basically).

1: I don't have any example, but I know is something I've encountered.

---

As an aside, digression, I was thinking about state management, and how we could have a Flux/Redux/NgRx like state datasource with a reactive interface that receives events, reduces the `(state+value+query) => state` and emits the new states… and is while thinking this that I though how a Query is so similar to the evens that use these state-management systems. That's how I came to the idea of removing the `value`. Actually, in these system the events are just plain objects with a `type` and a `payload`, and that's all it's needed to process them.
